plugins/tiles/src/main/java/org/apache/struts2/tiles/StrutsApplicationResource.java

```diff
@@ -23,14 +23,41 @@
 import java.io.File;
 import java.io.IOException;
 import java.io.InputStream;
+import java.net.MalformedURLException;
 import java.net.URL;
 
 public class StrutsApplicationResource extends PostfixedApplicationResource {
 
     private final URL url;
 
-    public StrutsApplicationResource(URL url) {
-        super(url.getPath());
+    /**
+     * workarounds WW-5011
+     */
+    private static String getExistedPath(URL url) throws MalformedURLException {
+        String path;
+        try {
+            //workarounds WW-5011 because includes ref in path
+            path = url.toURI().getPath();
+        } catch (Exception e) {
+            // fallback solution
+            path = url.getPath();
+
+            if (url.getRef() != null && !new File(path).exists()) {
+                // it's like WW-5011
+                path += "#" + url.getRef();
+            }
+        }
+
+        if (!new File(path).exists()) {
```

Comment:
(_Oops ... comment accidentally dropped from my review, re-adding as a single comment_)

The **File existence checks are dangerous** because the URL can be a non-file protocol resource.

For that reason _it is recommended that the two File existence checks are either removed_ or the logic changed so that _File existence checks are only performed after verifying the URL represents a file protocol resource_.  The check could be something like:
```java
if ("file".equalsIgnoreCase(url.getProtocol())) {
  // Perform file existence checks
}
// Return extracted path String (for non-File URLs or File URLs that pass existence checks).
```

Reply:
:+1: good idea, thanks!